### ISSUE TITLE
Weird behaviour with type variables on generic functions/methods returned by other generic functions

### ISSUE THREAD

**Issue by flisboac** [microsoft/pyright#4467]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**

Pyright doesn't seem to instantiate type variables on generic functions/methods returned by other generic functions.

I admit I don't truly understand what's happening, so categorizing this report as a bug may be a mistake. If so, I apologize.

**To Reproduce**

Please see the following snippet:

```python
from typing import Any, Callable, Generic, Iterable, TypeVar
from typing_extensions import Concatenate, ParamSpec, Final, TypeGuard


TIn = TypeVar("TIn")
TOut = TypeVar("TOut")
TOut_cov = TypeVar("TOut_cov", covariant=True)
PArgs = ParamSpec("PArgs")


def is_str(value: Any) -> TypeGuard[str]:
    return isinstance(value, str)


class FixerFunctionPipeStep(
    Generic[TOut_cov, PArgs],
):
    def __init__(
        self,
        impl: Callable[Concatenate[Iterable[Any], PArgs], Iterable[TOut_cov]],
        *args: PArgs.args,
        **kwargs: PArgs.kwargs,
    ) -> None:
        self.__impl: Final = impl
        self.__args: Final = args
        self.__kwargs: Final = kwargs

    def process(self, source: Iterable[Any]) -> Iterable[TOut_cov]:
        return self.__impl(source, *self.__args, **self.__kwargs)

    def __ror__(self, source: Iterable[Any]) -> Iterable[TOut_cov]:
        return self.process(source)


class FixerFunctionPipeStepFactory(Generic[TOut_cov, PArgs]):
    def __init__(
        self,
        impl: Callable[Concatenate[Iterable[Any], PArgs], Iterable[TOut_cov]],
    ) -> None:
        self.__impl = impl

    @classmethod
    def noargs(
        cls,
        impl: Callable[Concatenate[Iterable[Any], PArgs], Iterable[TOut_cov]],
    ) -> FixerFunctionPipeStep[TOut_cov, PArgs]:
        return cls(impl)()

    def __call__(
        self,
        *args: PArgs.args,
        **kwargs: PArgs.kwargs,
    ) -> FixerFunctionPipeStep[TOut_cov, PArgs]:
        return FixerFunctionPipeStep(
            self.__impl,
            *args,
            **kwargs,
        )


def _typed_as(
    input: Iterable[TIn],
    check: Callable[[TIn], TypeGuard[TOut]],
) -> Iterable[TOut]:
    for element in input:
        if check(element):
            yield element


typed_as: Final = FixerFunctionPipeStepFactory(_typed_as)


# Exhibit 1:
# - MYPY    reveals :: Revealed type is "test.FixerFunctionPipeStepFactory[TOut`-2, [check: def (TIn`-1) -> TypeGuard[TOut`-2]]]"
# - PYLANCE reveals :: Type of "typed_as" is "FixerFunctionPipeStepFactory[TOut@_typed_as, (check: (TIn@_typed_as) -> TypeGuard[TOut@_typed_as])]"
reveal_type(typed_as)

# Exhibit 2:
# - MYPY    reveals :: Revealed type is "test.FixerFunctionPipeStep[builtins.str, [check: def (Any) -> TypeGuard[builtins.str]]]"
# - PYLANCE reveals :: Type of "typed_as(is_str)" is "FixerFunctionPipeStep[TOut@_typed_as, (check: (TIn@_typed_as) -> TypeGuard[TOut@_typed_as])]"
reveal_type(typed_as(is_str))

# Exhibit 3:
# - MYPY    reveals :: Revealed type is "typing.Iterable[builtins.str]"
# - PYLANCE reveals :: Type of "typed_as(is_str).process((0, 1, 'a', {}))" is "Iterable[TOut@_typed_as]"
reveal_type(typed_as(is_str).process((0, 1, 'a', {})))

# Exhibit 4:
# - MYPY    reveals :: Revealed type is "typing.Iterable[builtins.str]"
# - PYLANCE reveals :: Type of "(0, 1, 'a', {}) | typed_as(is_str)" is "Iterable[TOut@_typed_as]"
reveal_type((0, 1, 'a', {}) | typed_as(is_str))

# Exhibit 5:
# - MYPY    reveals :: Revealed type is "builtins.list[builtins.str]"
# - PYLANCE reveals :: "list((0, 1, 'a', {}) | typed_as(is_str))" is "list[TOut@_typed_as]"
reveal_type(list((0, 1, 'a', {}) | typed_as(is_str)))

# Exhibit 6:
# - MYPY    reveals :: Revealed type is "builtins.list[builtins.str]"
# - PYLANCE reveals :: Type of "filtered_list" is "list[TOut@_typed_as]"
filtered_list = list((0, 1, 'a', {}) | typed_as(is_str))
reveal_type(filtered_list)
```

That last exhibit is what I expect, at the end of everything. But Pylance was getting it wrong, so I started to investigate what types both Pyright and Mypy reveal, expression by expression.

But first, some context. This is an effort to implement the famous [`pipe`](https://pypi.org/project/pipe/) library, but in a fully typed, and type-safe, manner (as much as possible, at least).

`FixerFunctionPipeStepFactory` is a generic class, and a factory for another generic class. The factory itself receives an external, generic/type-parameterized function, and when `__call__` is performed, an instance of `FixerFunctionPipeStep` is returned. `__call__` partially parameterizes the external function, and effectively creates a somewhat curried function.

Upon construction, `FixerFunctionPipeStep` contains all the necessary arguments for calling the original external function, except for the first argument (hence why the `Concatenate` there), which will be provided when `process` is called.

However, as shown in the exhibits, Pylance/Pyright do not instantiate/substitute type variables.

**Expected behavior**

I'd expect Pylance/Pyright behaviour to be exactly the same as what mypy exhibits.

My understanding may be lacking, but there's no reason for Pyright to not be able to deduce that the result of `process` is `Iterable[str]`, in all usages of `FixerFunctionPipeStep.process`.

**Screenshots or Code**

_Already shared previously._

Libraries/tools used:

- `python == 3.7.9`
- `typing-extensions == 4.4.0`
- `mypy == 0.991`

**VS Code extension or command-line**

Using Pylance v2023.1.10 on VSCode, on Windows (and NOT in WSL).

**Additional context**

_None._

**Comment by erictraut** [comment]
Mypy's behavior here is questionable, and I think pyright's is more correct. However, it's debatable because this code is relying on a portion of the type system that is under-specified.

PEP 484 indicates that type variables must be associated with either a class, function, or type alias. Type variables can be explicitly specialized (using subscript notation for a generic class or type alias) or "solved" in the context of a call expression that invokes a generic function or the constructor of a generic class. Type variables must be solved within the context of their associated scope. They don't have any meaning beyond this scope. If a type variable is not solved within the context of a call expression and is no longer in scope, it effectively remains unsolved with no way to solve it.

Mypy and pyright do make a special exception for this behavior in the case of a simple `Callable` return type. If a function returns a `Callable` that has unsolved type variables from another context, both mypy and pyright allow them to be solved in the context of the returned `Callable` under certain circumstances. Here's an example:

```python
from typing import Any, Callable, TypeVar

T = TypeVar("T")

def identity(x: T) -> T:
    return x
TCallable = TypeVar("TCallable", bound=Callable[..., Any])

def decorator1(x: TCallable) -> TCallable:
    return x

decorated1 = decorator1(identity)
reveal_type(decorated1)  # (x: T@identity) -> T@identity

d1_1 = decorated1(1)
reveal_type(d1_1)  # int
d1_2 = decorated1("")
reveal_type(d1_2)  # str
```

Pyright also allows the following variant, but mypy does not.

```python
def decorator2(x: Callable[[T], T]) -> Callable[[T], T]:
    return x

decorated2 = decorator2(identity)
reveal_type(decorated2)  # (T@identity) -> T@identity

d2_1 = decorated2(1)  # mypy: error: Argument 1 has incompatible type "int"; expected "T"
d2_2 = decorated2("")  # mypy: error; Argument 1 has incompatible type "str"; expected "T"
```

Neither pyright nor mypy extend the special case to callback protocols.

```python
class Transform(Protocol[T]):
    def __call__(self, __i: T) -> T:
        ...

def decorator3(x: Transform[T]) -> Transform[T]:
    return x

decorated3 = decorator3(identity)
d3_1 = decorated3(1)  # mypy & pyright: error
d3_2 = decorated3("")  # mypy & pyright: error
```

With all of that background in mind, let's look at the code sample posted in this issue.

In your examples, the `typed_as` variable is initialized with a value that comes from a call to a constructor of the generic class `FixerFunctionPipeStepFactory`. The type of the object that results from this constructor call is `FixerFunctionPipeStepFactory[TOut@_typed_as, (check: (TIn@_typed_as) -> TypeGuard[TOut@_typed_as])]`. These type variables are scoped to the `_typed_as` function. They have gone out of scope, so they can no longer be solved.

As evidence that you're doing something that's a bit amiss here, I challenge you to provide a proper type annotation for the variable `typed_as`. There's no way to write such a type annotation because doing so would require the use of unbound type variables.

I don't know why mypy is allowing these out-of-scope unsolved type variables to be solved in this context. This might be another special case that someone has added. More likely, it's unintentional behavior and a bug.

**Comment by flisboac** [comment]
I admit this use case is very sketchy. 😄 

The goal of `typed_as` is to be a factory callable for some other object implementing `process`, `__ror__`, etc. In my mind, it could be typed as such (or something along those lines, considering here all possible parameters, and removing the `ParamSpec` type-var for clarity):

```py
typed_as: Callable[
  [
    Callable[[Any], TypeGuard[TOut]],  # Receives a type guard...
  ],
  FixerFunctionPipeStep[TOut],  # And returns a parameterized instance of `FixerFunctionPipeStep`, for the guarded type.
]
```

This is what `FixerFunctionPipeStepFactory` was meant to be (ie. to decorate callables, so that they're substituted by  another with that signature).

`TOut` would fix the type for the elements in the resulting `Iterable` generated by `process`, etc. It should be bound at the time of the call to `typed_as` because that variable was substituted (contextually), and is explicitly passed along at class-level all the way to `FixerFunctionPipeStep`.

But given the context you provided, full resolution can only happen by means of `PArgs`, with knowledge only Pyright/Mypy would fully know, AND in a later moment (if ever). So, I guess it is not possible to solve `TOut` at all (ie. all the way) in this scenario, given how the type system is specified.

Having said that, is there any way to implement this pattern?

**Comment by erictraut** [comment]
I don't see a way to implement this pattern that is compatible with static typing.
